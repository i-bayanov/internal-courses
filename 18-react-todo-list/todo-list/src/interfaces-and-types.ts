export interface ITodoItem {
  id: string;
  title: string;
  completed: boolean;
}

export interface IAction {
  type: string;
  payload?: {
    todo?: ITodoItem;
    id?: string;
    value?: string;
    toggler?: boolean;
  };
}
