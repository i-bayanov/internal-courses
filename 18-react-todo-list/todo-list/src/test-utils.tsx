import React from 'react';
import { render } from 'react-dom';
import { BrowserRouter } from 'react-router-dom';

import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';

import todoReducer from './reducer';

const STORE = [
  {
    id: '1651234007958',
    title: 'Write a to-do list',
    completed: true,
  },
  {
    id: '1651234009086',
    title: 'Mark the first item as completed',
    completed: true,
  },
  {
    id: '1651234010110',
    title: '-',
    completed: false,
  },
  {
    id: '1651234864772',
    title: 'Come up with an interesting third point',
    completed: false,
  },
];

const mockStore = configureStore({ reducer: todoReducer, preloadedState: STORE });

// eslint-disable-next-line no-undef
const testRender = (component: JSX.Element, container: Element) => render(
  <Provider store={mockStore}>
    <BrowserRouter>
      {component}
    </BrowserRouter>
  </Provider>,
  container,
);

export { STORE, testRender };
