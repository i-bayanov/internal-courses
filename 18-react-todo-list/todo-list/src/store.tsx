import { configureStore } from '@reduxjs/toolkit';
import { IAction } from './interfaces-and-types';
import todoReducer from './reducer';

const store = configureStore({ reducer: todoReducer });

const dispatchAndSave = (action: IAction) => {
  store.dispatch(action);
  localStorage.setItem('todos', JSON.stringify(store.getState()));
};

type RootState = ReturnType<typeof store.getState>;

export { store, dispatchAndSave, RootState };
