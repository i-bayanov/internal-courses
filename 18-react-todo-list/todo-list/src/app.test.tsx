/* eslint-disable testing-library/no-node-access */
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { BrowserRouter } from 'react-router-dom';

import { configureStore } from '@reduxjs/toolkit';

import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Provider } from 'react-redux';
import { STORE, testRender } from './test-utils';
import todoReducer from './reducer';
import * as store from './store';
import App from './app';
import { IAction } from './interfaces-and-types';

let container: Element | null = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container!);
  container!.remove();
  container = null;
});

describe('Main component', () => {
  it('should renders correctly', () => {
    testRender(<App />, container!);
    expect(container).toMatchSnapshot();
  });
});

describe('Let\'s act with todo list and add some todos, mark them as completed, delete one etc.', () => {
  const mockStore = configureStore({ reducer: todoReducer, preloadedState: STORE });

  beforeEach(() => {
    // eslint-disable-next-line testing-library/no-render-in-setup
    render((
      <Provider store={mockStore}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    ));

    const spyDispatchAndSave = jest.spyOn(store, 'dispatchAndSave');
    spyDispatchAndSave.mockImplementation((action: IAction) => {
      mockStore.dispatch(action);
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should add "New todo"', () => {
    const input = screen.getByPlaceholderText('What needs to be done?') as HTMLInputElement;
    input.value = 'New todo';
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

    expect(screen.getAllByRole('listitem').filter((item) => item.dataset.id)).toHaveLength(5);
  });

  it('should delete todo with label "-"', () => {
    const todoToDelete = screen.getByText('-');
    const destroyButton = todoToDelete.nextElementSibling as HTMLButtonElement;
    fireEvent.pointerDown(destroyButton);

    expect(screen.getAllByRole('listitem').filter((item) => item.dataset.id)).toHaveLength(4);
  });

  it('should edit todo', () => {
    const todoToEdit = screen.getByText('Come up with an interesting third point');
    fireEvent.doubleClick(todoToEdit);
    const input = screen.getByDisplayValue('Come up with an interesting third point') as HTMLInputElement;
    input.value = 'Come up with the usual third point';
    fireEvent.blur(input);

    expect(screen.getByText('Come up with the usual third point')).toBeInTheDocument();
  });

  it('should mark todo as completed', () => {
    const todoToToggle = screen.getByText('Come up with the usual third point');
    const toggler = todoToToggle.previousElementSibling as HTMLInputElement;
    fireEvent.click(toggler);

    expect(screen.getByText(/left/i).textContent).toBe('1 item left');
  });

  it('should display only incompleted todos if "Active" filter is pressed', () => {
    const activeFilterButton = screen.getByText('Active');
    fireEvent.click(activeFilterButton);

    expect(screen.getAllByRole('listitem').filter((item) => item.dataset.id)).toHaveLength(1);
    expect(window.location.pathname).toBe('/active');
  });

  it('should display only completed todos if "Completed" filter is pressed', () => {
    const completedFilterButton = screen.getByText('Completed');
    fireEvent.click(completedFilterButton);

    expect(screen.getAllByRole('listitem').filter((item) => item.dataset.id)).toHaveLength(3);
    expect(window.location.pathname).toBe('/completed');
  });

  it('should display all todos if "All" filter is pressed', () => {
    const allFilterButton = screen.getByText('All');
    fireEvent.click(allFilterButton);

    expect(screen.getAllByRole('listitem').filter((item) => item.dataset.id)).toHaveLength(4);
    expect(window.location.pathname).toBe('/');
  });

  it('should mark all todos as completed', () => {
    const toggler = screen.getByLabelText('Mark all as completed');
    fireEvent.click(toggler);

    expect(screen.getByText(/left/i).textContent).toBe('0 items left');
  });

  it('should mark all todos as incompleted', () => {
    const toggler = screen.getByLabelText('Mark all as completed');
    fireEvent.click(toggler);

    expect(screen.getByText(/left/i).textContent).toBe('4 items left');
  });

  it('should mark all todos as completed and clear list', () => {
    const toggler = screen.getByLabelText('Mark all as completed');
    const clearButton = screen.getByText('Clear completed');
    fireEvent.click(toggler);
    fireEvent.pointerDown(clearButton);

    expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
  });
});
