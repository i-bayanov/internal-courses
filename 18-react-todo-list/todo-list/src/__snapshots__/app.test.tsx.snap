// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component should renders correctly 1`] = `
<div>
  <section
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as completed
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="completed "
          data-id="1651234007958"
        >
          <div
            class="view"
          >
            <input
              checked=""
              class="toggle"
              type="checkbox"
            />
            <label>
              Write a to-do list
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
        <li
          class="completed "
          data-id="1651234009086"
        >
          <div
            class="view"
          >
            <input
              checked=""
              class="toggle"
              type="checkbox"
            />
            <label>
              Mark the first item as completed
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
        <li
          class=" "
          data-id="1651234010110"
        >
          <div
            class="view"
          >
            <input
              class="toggle"
              type="checkbox"
            />
            <label>
              -
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
        <li
          class=" "
          data-id="1651234864772"
        >
          <div
            class="view"
          >
            <input
              class="toggle"
              type="checkbox"
            />
            <label>
              Come up with an interesting third point
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
      </ul>
      <footer
        class="footer"
      >
        <span
          class="todo-count"
        >
          2
           item
          s
           left
        </span>
        <ul
          class="filters"
        >
          <li>
            <a
              aria-current="page"
              class="selected"
              href="/"
            >
              All
            </a>
          </li>
          <li>
            <a
              class=""
              href="/active"
            >
              Active
            </a>
          </li>
          <li>
            <a
              class=""
              href="/completed"
            >
              Completed
            </a>
          </li>
        </ul>
        <button
          class="clear-completed"
          style="display: block;"
        >
          Clear completed
        </button>
      </footer>
    </section>
  </section>
</div>
`;
